components/common/src/command/package/install.rs

```diff
@@ -502,43 +625,139 @@ impl<'a> InstallTask<'a> {
 
     /// Adapter function to retrieve an installed package given an
     /// identifier, if it exists.
-    fn installed_package(&self, ident: &PackageIdent) -> Option<PackageInstall> {
-        PackageInstall::load(ident, Some(self.fs_root_path)).ok()
+    fn installed_package(&self, ident: &FullyQualifiedPackageIdent) -> Option<PackageInstall> {
+        PackageInstall::load(ident.as_ref(), Some(self.fs_root_path)).ok()
+    }
+
+    /// Checks for the latest installed package or cached artifact that matches a given package
+    /// identifier and returns a fully qualified package identifer if a match exists.
+    fn latest_installed_or_cached(
+        &self,
+        ident: &PackageIdent,
+    ) -> Result<FullyQualifiedPackageIdent> {
+        let latest_installed = self.latest_installed_ident(&ident);
+        let latest_cached = self.latest_cached_ident(&ident);
+        debug!(
+                "latest installed: {:?}, latest_cached: {:?}",
+                &latest_installed,
+                &latest_cached,
+            );
+        let latest = match (latest_installed, latest_cached) {
+            (Ok(pkg_install), Err(_)) => pkg_install,
+            (Err(_), Ok(pkg_artifact)) => pkg_artifact,
+            (Ok(pkg_install), Ok(pkg_artifact)) => {
+                if pkg_install.as_ref() > pkg_artifact.as_ref() {
+                    pkg_install
+                } else {
+                    pkg_artifact
+                }
+            }
+            (Err(_), Err(_)) => return Err(Error::PackageNotFound),
+        };
+        debug!("offline mode: winner: {:?}", &latest);
+
+        Ok(latest)
     }
 
-    // TODO (CM): This could return a plain bool IF we could ensure
-    // above that the package identifier is FULLY QUALIFIED
-    fn is_artifact_cached(&self, ident: &PackageIdent) -> Result<bool> {
-        Ok(self.cached_artifact_path(ident)?.is_file())
+    fn latest_installed_ident(&self, ident: &PackageIdent) -> Result<FullyQualifiedPackageIdent> {
```

Comment:
Since we're not using the `Err` for anything that I can see (nor `?` shortcuts), would it be more evocative to make the return an `Option` rather than a `Result`? To me the answer to the question `latest_installed_ident` may well be `None`, but I don't see that as an error _per se_.

If this would be too disruptive, it's not a big deal, but I mention it because I think there is value in the distinction between something that may be absent in normal behavior versus an error state.

Reply:
The `match` in the method is where we're handling any `Err`s from `PackageInstall::load()` (which returns a `Result<PackageInstall>`).

Reply:
Right, I'm wasn't suggesting changing the return type of `PackageInstall::load`, just changing the return type of `latest_installed_ident`. I tried it out a bit and I don't think it's actually much of a win. And being consistent with always using `Result<PackageInstall>` is probably best.